server/test/unit/com/thoughtworks/go/config/CachedGoConfigTest.java

```diff
@@ -19,62 +19,235 @@
+        assertThat(serverHealthService.getLogsAsText().contains("Invalid Merged Config"), is(true));
+        assertThat(serverHealthService.getLogsAsText().contains("pipe1"), is(true));
+    }
+
+    @Test
+    public void shouldUnSetErrorHealthStateWhenMergePasses() {
+        ConfigRepoConfig configRepo = configWatchList.getCurrentConfigRepos().get(0);
+        PartialConfig badPart = new PartialConfig(new PipelineGroups(
+                PipelineConfigMother.createGroup("part1",
+                        PipelineConfigMother.pipelineConfig("pipe1"), PipelineConfigMother.pipelineConfig("pipe1"))));
+        when(plugin.load(any(File.class), any(PartialConfigLoadContext.class))).thenReturn(badPart);
+
+        repoConfigDataSource.onCheckoutComplete(configRepo.getMaterialConfig(), folder, "321e");
+
+        assertThat(serverHealthService.getLogsAsText().contains("Invalid Merged Config"), is(true));
+
+        //fix partial
+        PartialConfig part1 = new PartialConfig(new PipelineGroups(
+                PipelineConfigMother.createGroup("part1", PipelineConfigMother.pipelineConfig("pipe1"))));
+        when(plugin.load(any(File.class), any(PartialConfigLoadContext.class))).thenReturn(part1);
+
+        repoConfigDataSource.onCheckoutComplete(configRepo.getMaterialConfig(), folder, "321e");
+
+        assertThat(serverHealthService.filterByScope(HealthStateScope.GLOBAL).isEmpty(), is(true));
+    }
+
+    @Test
+    public void shouldDelegateWritePipelineConfigCallToFileService() {
+        PipelineConfigService.SaveCommand saveCommand = mock(PipelineConfigService.SaveCommand.class);
+        GoFileConfigDataSource dataSource = mock(GoFileConfigDataSource.class);
+        CachedGoConfig cachedGoConfig = new CachedGoConfig(mock(ServerHealthService.class), dataSource);
+        PipelineConfig pipelineConfig = new PipelineConfig();
+        CachedGoConfig.PipelineConfigSaveResult saveResult = mock(CachedGoConfig.PipelineConfigSaveResult.class);
+        GoConfigHolder savedConfig = new GoConfigHolder(new BasicCruiseConfig(), new BasicCruiseConfig());
+        when(saveResult.getConfigHolder()).thenReturn(savedConfig);
+        GoConfigHolder holderBeforeUpdate = cachedGoConfig.loadConfigHolder();
+        Username user = new Username(new CaseInsensitiveString("user"));
+        when(dataSource.writePipelineWithLock(pipelineConfig, holderBeforeUpdate, saveCommand, user)).thenReturn(saveResult);
+
+        cachedGoConfig.writePipelineWithLock(pipelineConfig, saveCommand, user);
+        assertThat(cachedGoConfig.loadConfigHolder(), is(savedConfig));
+        assertThat(cachedGoConfig.currentConfig(), is(savedConfig.config));
+        assertThat(cachedGoConfig.loadForEditing(), is(savedConfig.configForEdit));
+        verify(dataSource).writePipelineWithLock(pipelineConfig, holderBeforeUpdate, saveCommand, user);
+    }
+
     protected CachedGoConfig cachedGoConfig;
```

Comment:
`cachedGoConfig` is always null here. 
Also it would be nice to put these fields higher.